dataset/slice.cpp

```diff
@@ -0,0 +1,101 @@
+#include "scipp/dataset/slice.h"
+#include "scipp/dataset/dataset.h"
+#include "scipp/units/dim.h"
+#include "scipp/variable/comparison.h"
+#include "scipp/variable/reduction.h"
+#include "scipp/variable/util.h"
+#include "scipp/variable/variable.h"
+
+namespace scipp::dataset {
+DataArrayConstView slice(const DataArrayConstView &to_slice, const Dim dim,
+                         const VariableConstView begin,
+                         const VariableConstView end) {
+  using namespace variable;
+  if (begin)
+    variable::detail::expect0D(begin.dims());
+  if (end)
+    variable::detail::expect0D(end.dims());
+  auto coords = to_slice.coords();
+  if (!coords.contains(dim))
+    throw except::DimensionNotFoundError(to_slice.dims(), dim);
+  auto [k, coord] = *coords.find(dim);
+  if (coord.dims().ndim() != 1) {
+    throw except::SizeError(
+        "Multi-dimensional coordinates not supported in slice");
+  }
+  const bool ascending = is_sorted(coord, dim, variable::SortOrder::Ascending);
+  const bool descending =
+      is_sorted(coord, dim, variable::SortOrder::Descending);
+
+  if (!(ascending ^ descending))
+    throw std::runtime_error("Coordinate must be monotomically increasing or "
+                             "decreasing for value slicing");
```

Comment:
For non-range slice of point coords I think the order does not and should not matter by spectrum number?
```python
spec = data['spectrum', sc.Variable(value=17)]
```
?